service/history/timerQueueProcessorBase.go

```diff
@@ -623,21 +624,21 @@ func (t *timerQueueProcessorBase) processArchiveHistoryEvent(task *persistence.T
 	} else if !ok {
 		return nil
 	}
-
 	req := &sysworkflow.ArchiveRequest{
 		DomainID:             task.DomainID,
 		WorkflowID:           task.WorkflowID,
 		RunID:                task.RunID,
 		EventStoreVersion:    msBuilder.GetEventStoreVersion(),
 		BranchToken:          msBuilder.GetCurrentBranch(),
-		LastFirstEventID:     msBuilder.GetLastFirstEventID(),
+		NextEventID:          msBuilder.GetNextEventID(),
 		CloseFailoverVersion: msBuilder.GetLastWriteVersion(),
 	}
+	key := definition.NewWorkflowIdentifier(task.DomainID, task.WorkflowID, task.RunID)
+	t.cache.Delete(key)
```

Comment:
I think a safer approach is to remove execution from cache when we call release.  How about we create a new overload for 'getOrCreateWorkflowExecution' which returns a release function that takes in removeFromCache flag.

Reply:
We can do that but it will result in somewhat messy code. There are currently a lot of places that call getOrCreateWorkflowExecutionWithTimeout. That is the method that actually creates the release function. To make this work you need to do one of the following not so nice things:

1. Duplicate all the logic that method into a new method with a different name which returns a release function that deletes from the cache when released.
2. Pass through a boolean everywhere that invokes this method then branch on this boolean to determine if you should delete from cache in release function.

I do not like either of these options from a code clarity perspective. Can you elaborate on the safety concern you have with the current approach.

Reply:
I made the change but I do not like how the code looks. There are basically three ways I could think of to implement this

1. Add new versions of the existing methods that know how to delete when released. I don't like this because it creates some really weird method names (this is the option I coded).
2. Add parameter to existing methods and check this parameter to know if the release function should delete from cache once invoked. This is not great because these methods are called in 43 places and only in once case do we actually want to delete from cache. So passing false in 42 places does not feel right.
3. Use the former approach of just deleting directly from the cache while holding the lock. You mentioned a safety concern around this but I am not sure I see the safety concern. Can we chat about this more because I like the former approach the best.

Reply:
I'm getting a little worried as I think about this more.  Our lru cache also has refCounting semantics a mutableStateCache (historyCache) heavily relies on making sure we use the same workflowExecutionContext for various correctness reasons.  For pinned caches, deleting items from the cache is actually very risky as it could lead to multiple objects representing the same key pinned by various api calls.  
Now thinking about this more, maybe we could just call workflowExecutionContext.clear() to remove mutable state rather than deleting item from the cache.  Sorry for not mentioning this earlier.  Let's discuss this more in person in the morning.